### PR TITLE
list layout 

### PR DESCRIPTION
**Pull request by DerThorsten**
* ready  for some initial review
* Implementation is done directly inside the classes (will change this later on with the next commits)

### CODE REVIEW

File: test/array_data_creation.hpp [hunk 1]
```diff
@@ -213,4 +215,87 @@ namespace sparrow::test
...
+
+        sparrow::array_data::buffer_type offset_buffer(sizeof(std::int64_t) * (values.size() + 1), 0);
+        auto offset_buffer_ptr = offset_buffer.data<std::int64_t>();
+        std::copy(offsets.begin(), offsets.end(), offset_buffer_ptr);
+        return offset_buffer;
+    }
+
+    template<class T>
+    sparrow::array_data make_array_data_for_list_of_scalars(
+        const std::vector<std::vector<T>>& values)
```
[the remark this answers is not in the document]
Comment (DerThorsten) [reply test/array_data_creation.hpp:235]: its in the tests...why making it complicated

File: test/test_list_layout.cpp [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+        for (auto& e : v) {
+            flat_values.push_back(e);
+        }
+    }
+    return flat_values;
+}
+
+// flatten tree nested vectors
+template <class T>
+std::vector<T> flatten_tree(std::vector<std::vector<std::vector<T>>> values)
```
[the remark this answers is not in the document]
Comment (DerThorsten) [reply test/test_list_layout.cpp:48]: its a test, no need for generic stuff here...

[the remark this answers is not in the document]
Comment (DerThorsten) [reply test/test_list_layout.cpp]: actually this unused dead code

File: test/array_data_creation.hpp [hunk 3]
```diff
@@ -213,4 +215,87 @@ namespace sparrow::test
     }   
 
 
+    template<class C>
+    sparrow::array_data::buffer_type build_offsets(C && values)
```
[the remark this answers is not in the document]
Comment (DerThorsten) [reply test/array_data_creation.hpp:219]: I can add this, but I think for tests there is no need whatsoever for concepts in test-code

File: test/test_list_layout.cpp [hunk 4]
```diff
@@ -0,0 +1,279 @@
...
+{
+   
+TEST_SUITE("list_layout")
+{
+    // this test will be invoked for each **scalar** type.
+    // Since this test uses  a "fixed_size_layout" as the inner layout, 
+    // it will **not** work for non-scalar types (ie std::string)
+    TEST_CASE_TEMPLATE_DEFINE("generic-scalar-test", T, generic_scalar_test)
+    {
+        SUBCASE("list<int>")
```
Comment (Alex-PLACET) [inline test/test_list_layout.cpp:40]:
```suggestion
        SUBCASE("list<T>")
```

File: test/test_list_layout.cpp [hunk 5]
```diff
@@ -0,0 +1,279 @@
...
+                for(std::size_t j = 0; j < values[i].size(); j++)
+                {
+                    auto maybe_value = list[j];
+                    CHECK_EQ(maybe_value.has_value(), true);
+                    CHECK_EQ(maybe_value.value(), values[i][j]);
+                }
+            }
+    
+        }
+        SUBCASE("list<list<int>>")
```
Comment (Alex-PLACET) [inline test/test_list_layout.cpp:75]:
```suggestion
        SUBCASE("list<list<T>>")
```

File: test/test_list_layout.cpp [hunk 6]
```diff
@@ -0,0 +1,279 @@
...
+                {
+                    outer_list_layout_type outer_list_layout(outer_list_array_data);
+                    CHECK_EQ(outer_list_layout.size(), values.size());
+                    for(auto mp : outer_list_layout.bitmap())
+                    {
+                        CHECK_EQ(bool(mp), true);
+                    }
+                }
+            }
+            SUBCASE("const operator[]")
```
Comment (Alex-PLACET) [inline test/test_list_layout.cpp:176]: Can you move this test next "operator[]"

File: test/test_list_layout.cpp [hunk 7]
```diff
@@ -0,0 +1,279 @@
...
+            {
+                SUBCASE("const")
+                {
+                    outer_list_layout_type outer_list_layout(outer_list_array_data);
+                    const outer_list_layout_type& const_outer_list_layout = outer_list_layout;
+                    for(auto list : const_outer_list_layout.values())
+                    {
+                        for(auto elemets  : list)
+                        {
+                            CHECK_EQ(elemets.has_value(), true);
```
Comment (Alex-PLACET) [inline test/test_list_layout.cpp:137]: Can you check the value too ?

File: test/test_list_layout.cpp [hunk 8]
```diff
@@ -0,0 +1,279 @@
...
+            auto list_array_data = test::make_array_data_for_list_of_scalars(values);
+
+            // auto flat_values = flatten(values);
+            using data_storage = sparrow::array_data;
+            using inner_layout_type = sparrow::fixed_size_layout<T, data_storage>;
+            using list_layout_type = sparrow::list_layout<inner_layout_type, data_storage,  std::int64_t>;
+            list_layout_type list_layout(list_array_data);
+
+
+            CHECK_EQ(list_layout.size(), values.size());
```
Comment (Alex-PLACET) [inline test/test_list_layout.cpp:59]:
```suggestion
            REQUIRE_EQ(list_layout.size(), values.size());
```

File: test/test_list_layout.cpp [hunk 9]
```diff
@@ -0,0 +1,279 @@
...
+            for(std::size_t i = 0; i < values.size(); i++)
+            {
+                auto maybe_list = list_layout[i];
+                CHECK_EQ(maybe_list.has_value(), true);
```
Comment (Alex-PLACET) [inline test/test_list_layout.cpp:63]:
```suggestion
                REQUIRE_EQ(maybe_list.has_value(), true);
```

File: test/test_list_layout.cpp [hunk 10]
```diff
@@ -0,0 +1,279 @@
...
+                auto list = maybe_list.value();
+                CHECK_EQ(list.size(), values[i].size());
```
Comment (Alex-PLACET) [inline test/test_list_layout.cpp:65]:
```suggestion
                REQUIRE_EQ(list.size(), values[i].size());
```

File: test/test_list_layout.cpp [hunk 11]
```diff
@@ -0,0 +1,279 @@
...
+                for(std::size_t j = 0; j < values[i].size(); j++)
+                {
+                    auto maybe_value = list[j];
+                    CHECK_EQ(maybe_value.has_value(), true);
```
Comment (Alex-PLACET) [inline test/test_list_layout.cpp:69]:
```suggestion
                    REQUIRE_EQ(maybe_value.has_value(), true);
```

File: test/test_list_layout.cpp [hunk 12]
```diff
@@ -0,0 +1,279 @@
...
+        
+            using data_storage = sparrow::array_data;
+            using inner_layout_type = sparrow::fixed_size_layout<T, data_storage>;
+            using inner_list_layout_type = sparrow::list_layout<inner_layout_type, data_storage,  std::int64_t>;
+            using outer_list_layout_type = sparrow::list_layout<inner_list_layout_type, data_storage,  std::int64_t>;
+            
+            SUBCASE("operator[]")
+            {
+                outer_list_layout_type outer_list_layout(outer_list_array_data);
+                CHECK_EQ(outer_list_layout.size(), values.size());
```
Comment (Alex-PLACET) [inline test/test_list_layout.cpp:105]:
```suggestion
                REQUIRE_EQ(outer_list_layout.size(), values.size());
```

File: test/test_list_layout.cpp [hunk 13]
```diff
@@ -0,0 +1,279 @@
...
+                for(std::size_t i = 0; i < values.size(); i++)
+                {
+                    auto maybe_list = outer_list_layout[i];
+                    CHECK_EQ(maybe_list.has_value(), true);
```
Comment (Alex-PLACET) [inline test/test_list_layout.cpp:109]:
```suggestion
                    REQUIRE_EQ(maybe_list.has_value(), true);
```

File: test/test_list_layout.cpp [hunk 14]
```diff
@@ -0,0 +1,279 @@
...
+                    auto list = maybe_list.value();
+                    CHECK_EQ(list.size(), values[i].size());
```
Comment (Alex-PLACET) [inline test/test_list_layout.cpp:111]:
```suggestion
                    REQUIRE_EQ(list.size(), values[i].size());
```

File: test/test_list_layout.cpp [hunk 15]
```diff
@@ -0,0 +1,279 @@
...
+                    for(std::size_t j = 0; j < values[i].size(); j++)
+                    {
+                        auto maybe_inner_list = list[j];
+                        CHECK_EQ(maybe_inner_list.has_value(), true);
```
Comment (Alex-PLACET) [inline test/test_list_layout.cpp:115]:
```suggestion
                        REQUIRE_EQ(maybe_inner_list.has_value(), true);
```

File: test/test_list_layout.cpp [hunk 16]
```diff
@@ -0,0 +1,279 @@
...
+                        auto inner_list = maybe_inner_list.value();
+                        CHECK_EQ(inner_list.size(), values[i][j].size());
```
Comment (Alex-PLACET) [inline test/test_list_layout.cpp:117]:
```suggestion
                        REQUIRE_EQ(inner_list.size(), values[i][j].size());
```

File: test/test_list_layout.cpp [hunk 17]
```diff
@@ -0,0 +1,279 @@
...
+                        for(std::size_t k = 0; k < values[i][j].size(); k++)
+                        {
+                            auto maybe_value = inner_list[k];
+                            CHECK_EQ(maybe_value.has_value(), true);
```
Comment (Alex-PLACET) [inline test/test_list_layout.cpp:121]:
```suggestion
                            REQUIRE_EQ(maybe_value.has_value(), true);
```

File: test/test_list_layout.cpp (hunk 7 above)
Comment (Alex-PLACET) [inline test/test_list_layout.cpp:137]:
```suggestion
                            REQUIRE_EQ(elements.has_value(), true);
```

File: test/test_list_layout.cpp [hunk 18]
```diff
@@ -0,0 +1,279 @@
...
+                        }
+                    }
+                }
+                SUBCASE("non-const")
+                {
+                    outer_list_layout_type outer_list_layout(outer_list_array_data);
+                    CHECK_EQ(outer_list_layout.size(), values.size());
+                    for(auto list : outer_list_layout.values())
+                    {
+                        for(auto elemets  : list)
```
Comment (Alex-PLACET) [inline test/test_list_layout.cpp:147]:
```suggestion
                        for(auto elements  : list)
```

File: test/test_list_layout.cpp [hunk 19]
```diff
@@ -0,0 +1,279 @@
...
+            }
+            SUBCASE("values")
+            {
+                SUBCASE("const")
+                {
+                    outer_list_layout_type outer_list_layout(outer_list_array_data);
+                    const outer_list_layout_type& const_outer_list_layout = outer_list_layout;
+                    for(auto list : const_outer_list_layout.values())
+                    {
+                        for(auto elemets  : list)
```
Comment (Alex-PLACET) [inline test/test_list_layout.cpp:135]:
```suggestion
                        for(const auto& elements  : list)
```

File: test/test_list_layout.cpp [hunk 20]
```diff
@@ -0,0 +1,279 @@
...
+                    {
+                        CHECK_EQ(bool(mp), true);
+                    }
+                }
+            }
+            SUBCASE("const operator[]")
+            {
+                outer_list_layout_type outer_list_layout(outer_list_array_data);
+                const outer_list_layout_type& const_outer_list_layout = outer_list_layout;
+                CHECK_EQ(const_outer_list_layout.size(), values.size());
```
Comment (Alex-PLACET) [inline test/test_list_layout.cpp:180]:
```suggestion
                REQUIRE_EQ(const_outer_list_layout.size(), values.size());
```

File: test/test_list_layout.cpp [hunk 21]
```diff
@@ -0,0 +1,279 @@
...
+                for(std::size_t i = 0; i < values.size(); i++)
+                {
+                    auto maybe_list = const_outer_list_layout[i];
+                    CHECK_EQ(maybe_list.has_value(), true);
```
Comment (Alex-PLACET) [inline test/test_list_layout.cpp:184]:
```suggestion
                    REQUIRE_EQ(maybe_list.has_value(), true);
```

File: test/test_list_layout.cpp [hunk 22]
```diff
@@ -0,0 +1,279 @@
...
+                    auto list = maybe_list.value();
+                    CHECK_EQ(list.size(), values[i].size());
```
Comment (Alex-PLACET) [inline test/test_list_layout.cpp:186]:
```suggestion
                    REQUIRE_EQ(list.size(), values[i].size());
```

File: test/test_list_layout.cpp [hunk 23]
```diff
@@ -0,0 +1,279 @@
...
+                    for(std::size_t j = 0; j < values[i].size(); j++)
+                    {
+                        auto maybe_inner_list = list[j];
+                        CHECK_EQ(maybe_inner_list.has_value(), true);
```
Comment (Alex-PLACET) [inline test/test_list_layout.cpp:190]:
```suggestion
                        REQUIRE_EQ(maybe_inner_list.has_value(), true);
```

File: test/test_list_layout.cpp [hunk 24]
```diff
@@ -0,0 +1,279 @@
...
+                        auto inner_list = maybe_inner_list.value();
+                        CHECK_EQ(inner_list.size(), values[i][j].size());
```
Comment (Alex-PLACET) [inline test/test_list_layout.cpp:192]:
```suggestion
                        REQUIRE_EQ(inner_list.size(), values[i][j].size());
```

File: test/test_list_layout.cpp [hunk 25]
```diff
@@ -0,0 +1,279 @@
...
+                        for(std::size_t k = 0; k < values[i][j].size(); k++)
+                        {
+                            auto maybe_value = inner_list[k];
+                            CHECK_EQ(maybe_value.has_value(), true);
```
Comment (Alex-PLACET) [inline test/test_list_layout.cpp:196]:
```suggestion
                            REQUIRE_EQ(maybe_value.has_value(), true);
```

File: test/test_list_layout.cpp [hunk 26]
```diff
@@ -0,0 +1,279 @@
...
+            }
+            SUBCASE("iterator")
+            {
+                outer_list_layout_type outer_list_layout(outer_list_array_data);
+                using difference_type = typename outer_list_layout_type::iterator::difference_type;
+                auto layout_iter  = outer_list_layout.begin();
+                for(std::size_t i = 0; i < values.size(); i++)
+                {
+                    auto maybe_list = layout_iter[static_cast<difference_type>(i)];
+                    CHECK_EQ(maybe_list.has_value(), true);
```
Comment (Alex-PLACET) [inline test/test_list_layout.cpp:210]:
```suggestion
                    REQUIRE_EQ(maybe_list.has_value(), true);
```

File: test/test_list_layout.cpp [hunk 27]
```diff
@@ -0,0 +1,279 @@
...
+                    auto list = maybe_list.value();
+                    auto iter = list.begin();
+                    CHECK_EQ(list.size(), values[i].size());
```
Comment (Alex-PLACET) [inline test/test_list_layout.cpp:213]:
```suggestion
                    REQUIRE_EQ(list.size(), values[i].size());
```

File: test/test_list_layout.cpp [hunk 28]
```diff
@@ -0,0 +1,279 @@
...
+                    for(std::size_t j = 0; j < values[i].size(); j++)
+                    {
+                        auto maybe_inner_list = iter[static_cast<difference_type>(j)];
+                        CHECK_EQ(maybe_inner_list.has_value(), true);
```
Comment (Alex-PLACET) [inline test/test_list_layout.cpp:217]:
```suggestion
                        REQUIRE_EQ(maybe_inner_list.has_value(), true);
```

File: test/test_list_layout.cpp [hunk 29]
```diff
@@ -0,0 +1,279 @@
...
+                        auto inner_list = maybe_inner_list.value();
+                        auto inner_iter = inner_list.begin();
+                        CHECK_EQ(inner_list.size(), values[i][j].size());
```
Comment (Alex-PLACET) [inline test/test_list_layout.cpp:220]:
```suggestion
                        REQUIRE_EQ(inner_list.size(), values[i][j].size());
```

File: test/test_list_layout.cpp [hunk 30]
```diff
@@ -0,0 +1,279 @@
...
+                        for(std::size_t k = 0; k < values[i][j].size(); k++)
+                        {
+                            auto maybe_value = inner_iter[static_cast<difference_type>(k)];
+                            CHECK_EQ(maybe_value.has_value(), true);
```
Comment (Alex-PLACET) [inline test/test_list_layout.cpp:224]:
```suggestion
                            REQUIRE_EQ(maybe_value.has_value(), true);
```

File: test/test_list_layout.cpp [hunk 31]
```diff
@@ -0,0 +1,279 @@
...
+            }
+            SUBCASE("const-iterator")
+            {
+                outer_list_layout_type outer_list_layout(outer_list_array_data);
+                using difference_type = typename outer_list_layout_type::iterator::difference_type;
+                auto layout_iter  = outer_list_layout.cbegin();
+                for(std::size_t i = 0; i < values.size(); i++)
+                {
+                    auto maybe_list = layout_iter[static_cast<difference_type>(i)];
+                    CHECK_EQ(maybe_list.has_value(), true);
```
Comment (Alex-PLACET) [inline test/test_list_layout.cpp:238]:
```suggestion
                    REQUIRE_EQ(maybe_list.has_value(), true);
```

File: test/test_list_layout.cpp [hunk 32]
```diff
@@ -0,0 +1,279 @@
...
+                    auto list = maybe_list.value();
+                    auto iter = list.cbegin();
+                    CHECK_EQ(list.size(), values[i].size());
```
Comment (Alex-PLACET) [inline test/test_list_layout.cpp:241]:
```suggestion
                    REQUIRE_EQ(list.size(), values[i].size());
```

File: test/test_list_layout.cpp [hunk 33]
```diff
@@ -0,0 +1,279 @@
...
+                    for(std::size_t j = 0; j < values[i].size(); j++)
+                    {
+                        auto maybe_inner_list = iter[static_cast<difference_type>(j)];
+                        CHECK_EQ(maybe_inner_list.has_value(), true);
```
Comment (Alex-PLACET) [inline test/test_list_layout.cpp:245]:
```suggestion
                        REQUIRE_EQ(maybe_inner_list.has_value(), true);
```

File: test/test_list_layout.cpp [hunk 34]
```diff
@@ -0,0 +1,279 @@
...
+                        auto inner_list = maybe_inner_list.value();
+                        auto inner_iter = inner_list.cbegin();
+                        CHECK_EQ(inner_list.size(), values[i][j].size());
```
Comment (Alex-PLACET) [inline test/test_list_layout.cpp:248]:
```suggestion
                        REQUIRE_EQ(inner_list.size(), values[i][j].size());
```

File: test/test_list_layout.cpp [hunk 35]
```diff
@@ -0,0 +1,279 @@
...
+                        for(std::size_t k = 0; k < values[i][j].size(); k++)
+                        {
+                            auto maybe_value = inner_iter[static_cast<difference_type>(k)];
+                            CHECK_EQ(maybe_value.has_value(), true);
```
Comment (Alex-PLACET) [inline test/test_list_layout.cpp:252]:
```suggestion
                            REQUIRE_EQ(maybe_value.has_value(), true);
```

File: test/test_typed_array_list_layout.cpp [hunk 36]
```diff
@@ -0,0 +1,60 @@
...
+        using inner_layout_type = sparrow::fixed_size_layout<int, data_storage>;
+        using list_layout_type = sparrow::list_layout<inner_layout_type, data_storage,  std::int64_t>;
+        using typed_array_type = typed_array<list_value_t<list_layout_type, true>, list_layout_type>;
+
+        typed_array_type array{list_array_data};
+
+        REQUIRE(array.size() == 3);
+        for(std::size_t i = 0; i < array.size(); ++i)
+        {   
+            CHECK(array[i].has_value());
```
Comment (Alex-PLACET) [inline test/test_typed_array_list_layout.cpp:48]:
```suggestion
            REQUIRE(array[i].has_value());
```

File: test/test_typed_array_list_layout.cpp [hunk 37]
```diff
@@ -0,0 +1,60 @@
...
+            auto value = array[i].value();
+            for(std::size_t j = 0; j <value.size(); ++j)
+            {
+                CHECK(value[j].has_value());
```
Comment (Alex-PLACET) [inline test/test_typed_array_list_layout.cpp:52]:
```suggestion
                REQUIRE(value[j].has_value());
```

File: test/test_list_layout.cpp (hunk 7 above)
Comment (DerThorsten) [reply test/test_list_layout.cpp:137]: lets not...lets rather write one **generic** consistency test for all layouts which checks of value() / bitmap() and normal iterators are consistent with each other.

**Review by Alex-PLACET — APPROVED** [review]

**Comment by SylvainCorlay** [comment]
🎉

**Pull request merged**